### ISSUE TITLE
Can I use SparseMatrix or HDF5Array

### ISSUE THREAD

**Issue by swbioinf** [RGLab/MAST#103]
Hi, 

I'm working with a large dataset (where a dense matrix is bad for RAM usage) so wondering if (and how) I can use a sparse matrix or HDF5Array with MAST? 

From what I can gather the conversion from singleCellExperiment to SingleCellAssay for use within the zlm() function is where I'm hitting problems. Am I going about this the right way (can I use SCE directly somehow, DIY a sparse matrix SCA, a different function e.t.c). 

I'm using v1.8.0, R version 3.5.0 and bioconductor 3.8

Thanks. 

-----------

Excample sce object
```
sampler <- c(0,0,0,0,0,0,1,2,2,2,4)
num_genes <- 50
num_cells <- 20 
gene_names=paste0("gene", 1:num_genes)
cell_names=paste0("cell", 1:num_cells)

condition <- factor(c(rep("Con",10), rep("test", 10)))
raw_counts <- replicate(num_cells, sample(sampler, num_genes, replace=TRUE))
log_counts <- log2( raw_counts + 1)

colnames(raw_counts) <-cell_names
rownames(raw_counts) <-gene_names
colnames(log_counts) <-cell_names
rownames(log_counts) <-gene_names

# With standard dense matrix
sce        <- SingleCellExperiment(assays=SimpleList(counts=raw_counts, logcounts=log_counts), colData=condition)

```

What I have tried sofar:
```

#-------------------------------
# Must convert to SCA to use zlm function?
zlm( ~condition, sca=sce) 
#Error in zlm(~condition, sca = sce) : 
#   `sca` must inherit from `data.frame` or `SingleCellAssay`

# Works just fine with dense matrix.
sca <- SceToSingleCellAssay(sce)
zlm( ~condition, sca=sca)

#-------------------------------
# Can I use a sparse matrix instead?

log_counts.sm <- Matrix::Matrix(raw_counts, sparse=TRUE) 
sce.sm        <- SingleCellExperiment(
   assays  = SimpleList(counts=raw_counts, logcounts=log_counts.sm),
   colData = condition)
sca.sm <- SceToSingleCellAssay(sce.sm)
#Error in (function (..., along = N, rev.along = NULL, new.names = NULL,  : 
#                       arg 'logcounts' is non-atomic


#-------------------------------
# What about via FromMatrix with a sparse matrix?
sca.fm <- FromMatrix(log_counts.sm, cData=condition)
#Error in checkArrayNames(exprsArray, cData, fData) : 
#   `exprsArray` must be numeric
## Yep : is.numeric(log_counts.sm) is FALSE

#-------------------------------
library(HDF5Array)
saveHDF5SummarizedExperiment(sce, dir="./scetest/", replace=TRUE)
sce_h5 <- loadHDF5SummarizedExperiment("./scetest/")

sca_h5 <- SceToSingleCellAssay(sce_h5)
#Error in (function (..., along = N, rev.along = NULL, new.names = NULL,  : 
#                       arg 'counts' is non-atomic

```

sessionInfo():
```
R version 3.5.0 (2018-04-23)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.3 LTS

Matrix products: default
BLAS: /mnt/software/apps/R/3.5.0/lib/R/lib/libRblas.so
LAPACK: /mnt/software/apps/R/3.5.0/lib/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C               LC_TIME=en_AU.UTF-8       
 [4] LC_COLLATE=en_AU.UTF-8     LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] HDF5Array_1.10.0            rhdf5_2.26.0               
 [3] MAST_1.8.0                  SingleCellExperiment_1.4.0 
 [5] SummarizedExperiment_1.12.0 DelayedArray_0.8.0         
 [7] BiocParallel_1.16.0         matrixStats_0.54.0         
 [9] Biobase_2.42.0              GenomicRanges_1.34.0       
[11] GenomeInfoDb_1.18.1         IRanges_2.16.0             
[13] S4Vectors_0.20.1            BiocGenerics_0.28.0        

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.0             bindr_0.1.1            compiler_3.5.0        
 [4] pillar_1.3.0           plyr_1.8.4             XVector_0.22.0        
 [7] prettyunits_1.0.2      progress_1.2.0         bitops_1.0-6          
[10] tools_3.5.0            zlibbioc_1.28.0        tibble_1.4.2          
[13] gtable_0.2.0           lattice_0.20-35        pkgconfig_2.0.2       
[16] rlang_0.3.0.1          Matrix_1.2-14          rstudioapi_0.8        
[19] yaml_2.2.0             bindrcpp_0.2.2         GenomeInfoDbData_1.2.0
[22] stringr_1.3.1          dplyr_0.7.8            hms_0.4.2             
[25] tidyselect_0.2.5       grid_3.5.0             glue_1.3.0            
[28] data.table_1.11.8      R6_2.3.0               Rhdf5lib_1.4.0        
[31] reshape2_1.4.3         purrr_0.2.5            ggplot2_3.1.0         
[34] magrittr_1.5           scales_1.0.0           assertthat_0.2.0      
[37] abind_1.4-5            colorspace_1.3-2       stringi_1.2.4         
[40] RCurl_1.95-4.11        lazyeval_0.2.1         munsell_0.5.0         
[43] crayon_1.3.4  
```

And thanks for development of this useful package :)

**Comment by gfinak** [comment]
@swbioinf sorry for the delayed response. 
Support for HDF5 and other back-ends is something we are actively working on, but it is not officially supported yet. 
Can you tell us more about your use case?

**Comment by amcdavid** [comment]
I **think** sparse matrices should almost work out of the box.  Let me do a little digging to see what the stumbling block might be.

**Comment by amcdavid** [comment]
As a workaround:

```
log_counts.sm = Matrix::Matrix(log_counts, sparse = TRUE) 
counts.sm = Matrix::Matrix(raw_counts, sparse = TRUE)
sce.sm        <- SingleCellExperiment(
    assays  = SimpleList(counts=counts.sm, logcounts=log_counts.sm),
    colData = condition)
sca = as(sce.sm, 'SingleCellAssay')
zlm( ~condition, sca=sca) 
```

This relies on there being an assay named with something in the list `magic_assay_names()` with will be assumed to contain logged (and maybe thresholded) counts, but doesn't actually attempt to check that this is true.  A more permanent fix:

-  Do not cast `FromMatrix(exprsArray)` into an array if it is not already an array.
-  Methods to gain an argument `exprs_values` like in `scater` to choose which assay to operate on.

**Comment by amcdavid** [comment]
Closed in  a1e6478
